core/server/master/src/main/java/alluxio/master/throttle/SystemMonitor.java

```diff
@@ -0,0 +1,391 @@
+    collectIndicators(pitTimeMS);
+    checkAndBackPressure();
+    mPrevPitInfo.setPitTime(pitTimeMS);
+    mPrevPitInfo.setSystemStatus(mCurrentSystemStatus);
+    mPrevPitInfo.setJVMPauseTime(mServerIndicatorsList
+        .get(mServerIndicatorsList.size() - 1).getPITTotalJVMPauseTimeMS());
+  }
+
+  private void collectIndicators(long pitTimeMS) {
+    collectServerIndicators(pitTimeMS);
+    collectFileSystemIndicators(pitTimeMS);
+  }
+
+  private void produceDeltaFilesystemIndicators() {
+    // Only be able to calculate the delta if the two filesystem indicators are fetched
+    // continuously, this means both previous and current status should be either STRESSED
+    // or OVERLOADED
+    if ((mPrevPitInfo.getSystemStatus() == SystemStatus.STRESSED
+        || mPrevPitInfo.getSystemStatus() == SystemStatus.OVERLOADED)
+        && (mCurrentSystemStatus == SystemStatus.STRESSED
+        || mCurrentSystemStatus == SystemStatus.OVERLOADED)
+        && mPrevFileSystemIndicators != null
+        && mCurrentFileSystemIndicators != null) {
+      mDeltaFilesystemIndicators = new FileSystemIndicator(mCurrentFileSystemIndicators);
+      mDeltaFilesystemIndicators.deltaTo(mPrevFileSystemIndicators);
+    } else {
+      mDeltaFilesystemIndicators = null;
+    }
+  }
+
+  private void collectFileSystemIndicators(long snapshotTimeMS) {
+    mPrevFileSystemIndicators = mCurrentFileSystemIndicators;
+    if (mCurrentSystemStatus == SystemStatus.IDLE
+        || mCurrentSystemStatus == SystemStatus.ACTIVE) {
+      // don't collect information in case of IDLE and ACTIVE
+      mCurrentFileSystemIndicators = null;
+      mDeltaFilesystemIndicators = null;
+      return;
+    }
+
+    FileSystemIndicator fileSystemIndicator = new FileSystemIndicator();
+    fileSystemIndicator.setPitTimeMS(snapshotTimeMS);
+
+    mCurrentFileSystemIndicators = fileSystemIndicator;
+    produceDeltaFilesystemIndicators();
+  }
+
+  private void collectServerIndicators(long snapshotTimeMS) {
+    ServerIndicator serverIndicator
+        = ServerIndicator.createFromMetrics(
+        mPrevPitInfo != null ? mPrevPitInfo.getJVMPauseTime() : 0);
+    serverIndicator.setPitTimeMS(snapshotTimeMS);
+
+    mServerIndicatorsList.add(serverIndicator);
+    if (mServerIndicatorsList.size() > mMaxNumberOfSnapshot && !mServerIndicatorsList.isEmpty()) {
+      ServerIndicator toBeRemoved = mServerIndicatorsList.get(0);
+      if (mAggregatedServerIndicators != null) {
+        mAggregatedServerIndicators.reduction(toBeRemoved);
+      }
+      mServerIndicatorsList.remove(0);
+    }
+
+    if (mAggregatedServerIndicators == null) {
+      mAggregatedServerIndicators = new ServerIndicator(serverIndicator);
+    } else {
+      mAggregatedServerIndicators.addition(serverIndicator);
+    }
+  }
+
+  private StatusTransition statusCheck(ServerIndicator aggregatedServerIndicator,
+      ServerIndicator pitServerIndicator,
+      ServerIndicator lowBoundaryPitIndicator,
```

Comment:
Can you use @Nullable annotation please, and indicate add a comment what it means when the input is null (i.e. if lowBoundPitIndicator is null then status will not deescalate)